geth/common/geth/node.go

```diff
@@ -0,0 +1,31 @@
+package geth
+
+import (
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/node"
+)
+
+// Node interface for geth.Node.
+type Node interface {
```

Comment:
I don't see how `Node` may be a useful interface. We can't create an interface which is exactly like a struct just because we want to mock it. This is called interface pollution and introduces only confusion.

Reply:
Structs has inner state that isn`t thread-safe. Node, RpcClient, Whisper - if we exposes them as structs we cannot control thread-safaty. So we could either incapsulate all them`s interfaces in NodeManager (even more getters/setters) or create interfaces and expose only thread safe interfaces.

Reply:
> Structs has inner state that isnt thread-safe.

That's why it should export only methods that are thread-safe (export by using a capital letter). I don't know how interface helps that.

Reply:
If we use interfaces instead of types themselfes we exposes only methods without inner state of object. And we do saparate responsobilities: instead of huge NodeManager interface with Node, Whisper, etc we have smaller NadeManager interface and smaller Node, Whisper etc interfaces.

Reply:
It can be done by not exporting struct names. Following this approach, we should create an interface for almost every struct but it's not an interface role.

Interfaces are used when there might be many implementations or the implementation is provided.

Reply:
Could you describe this more: we have thread unsafe methods in 3rd parties packages and we have exported fields in this structs, how can we do safe NodeManager without interfaces or huge extending NodeManager interface?

Reply:
Could you point me to such an unsafe method?

I am not sure I understand your question correctly, but let's say we want to use `geth.Node` which comes from go-ethereum and it exports methods that are not thread-safe. 

We can use composition and create our `Node`:
```go
type Node struct {
  mu sync.Mutex
  node *geth.Node
}

func (n *Node) Start() error {
  n.mu.Lock()
  defer n.mu.Unlock()

  return n.node.Start()
}
```

Then, we can have `Manager` that uses our `Node`:
```go
type Manager struct {
  node *Node
}
```

Of course, `Node` and `Manager` will be in the same package, so there is a chance that in `Manager`s method, one can write `m.node.node` but it should be detected during a code review and ensured by unit tests that both `Node` and `Manager` are thread-safe. 

Using interfaces for maintaining thread-safety is an illusion because it will work only as long as underlying structs are thread-safe.

Reply:
That is, in fact it all comes down to that for a pair of objects and mutexes to extract from the NodeManager. But if we compose them by type not interface we couldn't use mocks for them. And we'll made a huge amount of boilerplate code like:
```
func (n *Node) Start() error {
  n.mu.Lock()
  defer n.mu.Unlock()

  return n.node.Start()
}
```

Reply:
> But if we compose them by type not interface we couldn't use mocks for them.

As I said, creating interfaces just because we want to write mocks is a bad idea. Take a look here https://github.com/ethereum/go-ethereum/blob/master/node/node_test.go. They test `Node` without any mocks. Mocks are sometimes useful but not always.

> And we'll made a huge amount of boilerplate code like:

Not really, we don't use many `geth.Node` methods actually. I counted five in `manager.go`.

Reply:
As far as i see this cose uses stabs. This solution have its own pros and cons: not all can be tested with stabs.
Why we should prefer stabs against mocks?